File: csharp/UrakawaToolkit/media/ExternalAudioMedia.cs
using System;
using System.Xml;
using urakawa.media.timing;

namespace urakawa.media
{
	/// <summary>
	/// AudioMedia is the audio object.
	/// It is time-based and comes from an external source.
	/// </summary>
	public class ExternalAudioMedia : IAudioMedia, IClipped, ILocated
	{
		private string mSrc;
		private Time mClipBegin;
		private Time mClipEnd;
		private IMediaFactory mFactory;
		private string mLanguage;

		private void resetClipTimes()
		{
			mClipBegin = new Time();
			mClipEnd = new Time(TimeSpan.MaxValue);
			mSrc = "";
		}

		/// <summary>
		/// Constructor setting the associated <see cref="IMediaFactory"/>
		/// </summary>
		/// <param name="fact">The <see cref="IMediaFactory"/> with which to associate</param>
		protected internal ExternalAudioMedia(IMediaFactory fact)
		{
			if (fact == null)
			{
				throw new exception.MethodParameterIsNullException("Factory is null");
			}
			mFactory = fact;
			mSrc = "";
			resetClipTimes();
			mLanguage = null;
		}
		
		#region IMedia members

		/// <summary>
		/// Sets the language of the external audio media
		/// </summary>
		/// <param name="lang">The new language, can be null but not an empty string</param>
		public void setLanguage(string lang)
		{
			if (lang == "")
			{
				throw new exception.MethodParameterIsEmptyStringException(
					"The language can not be an empty string");
			}
			mLanguage = lang;
		}

		/// <summary>
		/// Gets the language of the external audio media
		/// </summary>
		/// <returns>The language</returns>
		public string getLanguage()
		{
			return mLanguage;
		}
		/// <summary>
		/// This always returns true, because
		/// audio media is always considered continuous
		/// </summary>
		/// <returns></returns>
		public bool isContinuous()
		{
			return true;
		}

		/// <summary>
		/// This always returns false, because
		/// audio media is never considered discrete
		/// </summary>
		/// <returns></returns>
		public bool isDiscrete()
		{
			return false;
		}

		/// <summary>
		/// This always returns false, because
		/// a single media object is never considered to be a sequence
		/// </summary>
		/// <returns></returns>
		public bool isSequence()
		{
			return false;
		}
		
		IMedia IMedia.copy()
		{
			return copy();
		}

		/// <summary>
		/// Copy function which returns an <see cref="IAudioMedia"/> object
		/// </summary>
		/// <returns>A copy of this</returns>
		/// <exception cref="exception.FactoryCannotCreateTypeException">
		/// Thrown when the <see cref="IMediaFactory"/> associated with this 
		/// can not create an <see cref="ExternalAudioMedia"/> matching the QName of <see cref="ExternalAudioMedia"/>
		/// </exception>
		public ExternalAudioMedia copy()
		{
			return export(getMediaFactory().getPresentation());
		}

		IMedia IMedia.export(Presentation destPres)
		{
			return export(destPres);
		}

		/// <summary>
		/// Exports the external audio media to a destination <see cref="Presentation"/>
		/// </summary>
		/// <param name="destPres">The destination presentation</param>
		/// <returns>The exported external audio media</returns>
		public ExternalAudioMedia export(Presentation destPres)
		{
			ExternalAudioMedia exported = destPres.getMediaFactory().createMedia(
				getXukLocalName(), getXukNamespaceUri()) as ExternalAudioMedia;
			if (exported==null)
			{
				throw new exception.FactoryCannotCreateTypeException(String.Format(
					"The MediaFacotry cannot create a ExternalAudioMedia matching QName {1}:{0}",
					getXukLocalName(), getXukNamespaceUri()));
			}
			exported.setClipBegin(getClipBegin().copy());
			exported.setClipEnd(getClipEnd().copy());
			exported.setSrc(getSrc());
			return exported;
		}

		/// <summary>
		/// Gets the <see cref="IMediaFactory"/> associated with <c>this</c>
		/// </summary>
		/// <returns>The <see cref="IMediaFactory"/></returns>
		public IMediaFactory getMediaFactory()
		{
			return mFactory;
		}

		#endregion

		#region Old IXUKAble members

		///// <summary>
		///// Fill in audio data from an XML source.
		///// Assume that the XmlReader cursor is at the opening audio tag.
		///// </summary>
		///// <param name="source">the input XML source</param>
		///// <returns>true or false, depending on whether the data could be processed</returns>
		//public bool XukIn(System.Xml.XmlReader source)
		//{
		//  if (source == null)
		//  {
		//    throw new exception.MethodParameterIsNullException("Xml Reader is null");
		//  }
		//  if (source.NodeType != System.Xml.XmlNodeType.Element) return false;

		//  string cb = source.GetAttribute("clipBegin");
		//  string ce = source.GetAttribute("clipEnd");

		//  resetClipTimes();

		//  try
		//  {
		//    Time ceTime = new Time(ce);
		//    Time cbTime = new Time(cb);
		//    if (cbTime.isNegativeTimeOffset())
		//    {
		//      setClipBegin(cbTime);
		//      setClipEnd(ceTime);
		//    }
		//    else
		//    {
		//      setClipEnd(ceTime);
		//      setClipBegin(cbTime);
		//    }
		//  }
		//  catch (exception.TimeStringRepresentationIsInvalidException)
		//  {
		//    return false;
		//  }
		//  catch (exception.MethodParameterIsOutOfBoundsException)
		//  {
		//    return false;
		//  }
		//  IMediaLocation loc = null;

		//  if (!source.IsEmptyElement)
		//  {
		//    while (source.Read())
		//    {
		//      if (source.NodeType == XmlNodeType.Element)
		//      {
		//        loc = getMediaFactory().createMediaLocation(source.LocalName, source.NamespaceURI);
		//        if (loc == null)
		//        {
		//          if (!source.IsEmptyElement)
		//          {
		//            //Read past unrecognized element
		//            source.ReadSubtree().Close();
		//          }
		//        }
		//        else
		//        {
		//          if (!loc.XukIn(source)) return false;
		//        }
		//      }
		//      else if (source.NodeType == XmlNodeType.EndElement)
		//      {
		//        break;
		//      }
		//      if (source.EOF) break;
		//    }
		//  }
		//  if (loc == null) return false;
		//  setSrc(loc);
		//  return true;
		//}
		#endregion


		#region IXUKAble members



		/// <summary>
		/// Reads the <see cref="ExternalAudioMedia"/> from a ExternalAudioMedia xuk element
		/// </summary>
		/// <param name="source">The source <see cref="System.Xml.XmlReader"/></param>
		public void XukIn(XmlReader source)
		{
			if (source == null)
			{
				throw new exception.MethodParameterIsNullException("Can not XukIn from an null source XmlReader");
			}
			if (source.NodeType != XmlNodeType.Element)
			{
				throw new exception.XukException("Can not read ExternalAudioMedia from a non-element node");
			}
			try
			{
				XukInAttributes(source);
				if (!source.IsEmptyElement)
				{
					while (source.Read())
					{
						if (source.NodeType == XmlNodeType.Element)
						{
							XukInChild(source);
						}
						else if (source.NodeType == XmlNodeType.EndElement)
						{
							break;
						}
						if (source.EOF) throw new exception.XukException("Unexpectedly reached EOF");
					}
				}

			}
			catch (exception.XukException e)
			{
				throw e;
			}
			catch (Exception e)
			{
				throw new exception.XukException(
					String.Format("An exception occured during XukIn of ExternalAudioMedia: {0}", e.Message),
					e);
			}
		}

		/// <summary>
		/// Reads the attributes of a ExternalAudioMedia xuk element.
		/// </summary>
		/// <param name="source">The source <see cref="XmlReader"/></param>
		protected virtual void XukInAttributes(XmlReader source)
		{
			resetClipTimes();
			Time cbTime, ceTime;
			try
			{
				cbTime = new Time(source.GetAttribute("clipBegin"));
			}
			catch (exception.CheckedException e)
			{
				throw new exception.XukException(
					String.Format("clipBegin attribute is missing or has invalid value: {0}", e.Message),
					e);
			}
			try
			{
				ceTime = new Time(source.GetAttribute("clipEnd"));
			}
			catch (exception.CheckedException e)
			{
				throw new exception.XukException(
					String.Format("clipEnd attribute is missing or has invalid value: {0}", e.Message),
					e);
			}
			if (cbTime.isNegativeTimeOffset())
			{
				setClipBegin(cbTime);
				setClipEnd(ceTime);
			}
			else
			{
				setClipEnd(ceTime);
				setClipBegin(cbTime);
			}
			string s = source.GetAttribute("src");
			if (s == null)
			{
				throw new exception.XukException(
					"src attribute is missing from ExternamAudioMedia element");
			}
			setSrc(s);
			string lang = source.GetAttribute("Language").Trim();
			if (lang != "") lang = null;
			setLanguage(lang);
		}

		/// <summary>
		/// Reads a child of a ImageMedia xuk element. 
		/// </summary>
		/// <param name="source">The source <see cref="XmlReader"/></param>
		protected virtual void XukInChild(XmlReader source)
		{
			bool readItem = false;
			if (source.NamespaceURI == ToolkitSettings.XUK_NS)
			{
				readItem = true;
				switch (source.LocalName)
				{
					default:
						readItem = false;
						break;
				}
			}
			if (!(readItem || source.IsEmptyElement))
			{
				source.ReadSubtree().Close();//Read past unknown child 
			}
		}

		/// <summary>
		/// Write a ExternalAudioMedia element to a XUK file representing the <see cref="ExternalAudioMedia"/> instance
		/// </summary>
		/// <param localName="destination">The destination <see cref="System.Xml.XmlWriter"/></param>
		public void XukOut(System.Xml.XmlWriter destination)
		{
			if (destination == null)
			{
				throw new exception.MethodParameterIsNullException(
					"Can not XukOut to a null XmlWriter");
			}

			try
			{
				destination.WriteStartElement(getXukLocalName(), getXukNamespaceUri());
				XukOutAttributes(destination);
				XukOutChildren(destination);
				destination.WriteEndElement();
			}
			catch (exception.XukException e)
			{
				throw e;
			}
			catch (Exception e)
			{
				throw new exception.XukException(
					String.Format("An exception occured during XukOut of ExternalAudioMedia: {0}", e.Message),
					e);
			}
		}

		/// <summary>
		/// Writes the attributes of a ExternalAudioMedia element
		/// </summary>
		/// <param name="destination">The destination <see cref="XmlWriter"/></param>
		protected virtual void XukOutAttributes(XmlWriter destination)
		{
			destination.WriteAttributeString("clipBegin", this.getClipBegin().ToString());
			destination.WriteAttributeString("clipEnd", this.getClipEnd().ToString());
			destination.WriteAttributeString("src", getSrc());
			if (getLanguage() != null) destination.WriteAttributeString("Language", getLanguage());
		}

		/// <summary>
		/// Write the child elements of a ExternalAudioMedia element.
		/// </summary>
		/// <param name="destination">The destination <see cref="XmlWriter"/></param>
		protected virtual void XukOutChildren(XmlWriter destination)
		{

		}

		
		/// <summary>
		/// Gets the local name part of the QName representing a <see cref="ExternalAudioMedia"/> in Xuk
		/// </summary>
		/// <returns>The local name part</returns>
		public string getXukLocalName()
		{
			return this.GetType().Name;
		}

		/// <summary>
		/// Gets the namespace uri part of the QName representing a <see cref="ExternalAudioMedia"/> in Xuk
		/// </summary>
		/// <returns>The namespace uri part</returns>
		public string getXukNamespaceUri()
		{
			return urakawa.ToolkitSettings.XUK_NS;
		}

		#endregion

		#region ILocated Members

		/// <summary>
		/// Gets the src of <c>this</c>
		/// </summary>
		/// <returns>The src</returns>
		public string getSrc()
		{
			return mSrc;
		}

		/// <summary>
		/// Sets the src of <c>this</c>
		/// </summary>
		/// <param name="src">The new src</param>
		/// <exception cref="exception.MethodParameterIsNullException">
		/// Thrown when <paramref name="src"/> is <c>null</c></exception>
		public void setSrc(string src)
		{
			if (src == null)
			{
				throw new exception.MethodParameterIsNullException("The src can not be null");
			}
			mSrc = src;
		}

		#endregion

		#region IContinuous Members

		/// <summary>
		/// Gets the duration of <c>this</c>
		/// </summary>
		/// <returns>A <see cref="TimeDelta"/> representing the duration</returns>
		public TimeDelta getDuration()
		{
			return getClipEnd().getTimeDelta(getClipBegin());
		}

		#endregion

		#region IClipped Members

		/// <summary>
		/// Gets the clip begin <see cref="Time"/> of <c>this</c>
		/// </summary>
		/// <returns>Clip begin</returns>
		public Time getClipBegin()
		{
			return mClipBegin;
		}

		/// <summary>
		/// Gets the clip end <see cref="Time"/> of <c>this</c>
		/// </summary>
		/// <returns>Clip end</returns>
		public Time getClipEnd()
		{
			return mClipEnd;
		}

		/// <summary>
		/// Sets the clip begin <see cref="Time"/>
		/// </summary>
		/// <param name="beginPoint">The new clip begin <see cref="Time"/></param>
		/// <exception cref="exception.MethodParameterIsNullException">
		/// Thrown when <paramref localName="beginPoint"/> is <c>null</c>
		/// </exception>
		/// <exception cref="exception.MethodParameterIsOutOfBoundsException">
		/// Thrown when <paramref localName="beginPoint"/> is beyond clip end of <c>this</c>
		/// </exception>
		public void setClipBegin(Time beginPoint)
		{
			if (beginPoint==null)
			{
				throw new exception.MethodParameterIsNullException("ClipBegin can not be null");
			}
			if (beginPoint.isGreaterThan(getClipEnd()))
			{
				throw new exception.MethodParameterIsOutOfBoundsException(
					"ClipBegin can not be after ClipEnd"); 
			}
			mClipBegin = beginPoint;
		}

		/// <summary>
		/// Sets the clip end <see cref="Time"/>
		/// </summary>
		/// <param name="endPoint">The new clip end <see cref="Time"/></param>
		/// <exception cref="exception.MethodParameterIsNullException">
		/// Thrown when <paramref localName="endPoint"/> is <c>null</c>
		/// </exception>
		/// <exception cref="exception.MethodParameterIsOutOfBoundsException">
		/// Thrown when <paramref localName="endPoint"/> is before clip begin of <c>this</c>
		/// </exception>
		public void setClipEnd(Time endPoint)
		{
			if (endPoint == null)
			{
				throw new exception.MethodParameterIsNullException("ClipEnd can not be null");
			}
			if (endPoint.isLessThan(getClipBegin()))
			{
				throw new exception.MethodParameterIsOutOfBoundsException(
					"ClipEnd can not be before ClipBegin");
			}
			mClipEnd = endPoint;
		}

		IContinuous IContinuous.split(Time splitPoint)
		{
			return split(splitPoint);
		}

		/// <summary>
		/// Splits <c>this</c> at a given <see cref="Time"/>
		/// </summary>
		/// <param name="splitPoint">The <see cref="Time"/> at which to split - 
		/// must be between clip begin and clip end <see cref="Time"/>s</param>
		/// <returns>
		/// A newly created <see cref="IAudioMedia"/> containing the audio after,
		/// <c>this</c> retains the audio before <paramref localName="splitPoint"/>.
		/// </returns>
		public ExternalAudioMedia split(Time splitPoint)
		{
			if (splitPoint==null)
			{
				throw new exception.MethodParameterIsNullException(
					"The time at which to split can not be null");
			}
			if (splitPoint.isLessThan(getClipBegin()))
			{
				throw new exception.MethodParameterIsOutOfBoundsException(
					"The split time can not be before ClipBegin");
			}
			if (splitPoint.isGreaterThan(getClipEnd()))
			{
				throw new exception.MethodParameterIsOutOfBoundsException(
					"The split time can not be after ClipEnd");
			}
			ExternalAudioMedia splitAM = (ExternalAudioMedia)copy();
			setClipEnd(splitPoint);
			splitAM.setClipBegin(splitPoint);
			return splitAM;

		}

		#endregion

		#region IValueEquatable<IMedia> Members

		/// <summary>
		/// Conpares <c>this</c> with a given other <see cref="IMedia"/> for equality
		/// </summary>
		/// <param name="other">The other <see cref="IMedia"/></param>
		/// <returns><c>true</c> if equal, otherwise <c>false</c></returns>
		public bool ValueEquals(IMedia other)
		{
			if (other == null) return false;
			if (getLanguage() != other.getLanguage()) return false;
			if (GetType() != other.GetType()) return false;
			ExternalAudioMedia otherAudio = (ExternalAudioMedia)other;
			if (getSrc()!=otherAudio.getSrc()) return false;
			if (!getClipBegin().isEqualTo(otherAudio.getClipBegin())) return false;
			if (!getClipEnd().isEqualTo(otherAudio.getClipEnd())) return false;
			return true;
		}

		#endregion

	}
}

File: trunk/java/core/src/org/daisy/urakawa/media/ExternalAudioMedia.java
package org.daisy.urakawa.media;

import java.net.URI;
import java.net.URISyntaxException;

import org.daisy.urakawa.FactoryCannotCreateTypeException;
import org.daisy.urakawa.Presentation;
import org.daisy.urakawa.events.DataModelChangedEvent;
import org.daisy.urakawa.events.Event;
import org.daisy.urakawa.events.EventHandler;
import org.daisy.urakawa.events.IEventHandler;
import org.daisy.urakawa.events.IEventListener;
import org.daisy.urakawa.events.media.ClipChangedEvent;
import org.daisy.urakawa.events.media.SrcChangedEvent;
import org.daisy.urakawa.exception.IsNotInitializedException;
import org.daisy.urakawa.exception.MethodParameterIsEmptyStringException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.media.timing.ITime;
import org.daisy.urakawa.media.timing.ITimeDelta;
import org.daisy.urakawa.media.timing.Time;
import org.daisy.urakawa.media.timing.TimeOffsetIsOutOfBoundsException;
import org.daisy.urakawa.media.timing.TimeStringRepresentationIsInvalidException;
import org.daisy.urakawa.nativeapi.IXmlDataReader;
import org.daisy.urakawa.nativeapi.IXmlDataWriter;
import org.daisy.urakawa.progress.IProgressHandler;
import org.daisy.urakawa.progress.ProgressCancelledException;
import org.daisy.urakawa.xuk.XukDeserializationFailedException;
import org.daisy.urakawa.xuk.XukSerializationFailedException;

/**
 * Reference implementation of the interface.
 * 
 * @leafInterface see {@link org.daisy.urakawa.LeafInterface}
 * @see org.daisy.urakawa.LeafInterface
 */
public class ExternalAudioMedia extends AbstractAudioMedia implements ILocated,
        IClipped
{
    // BEGIN ILocated common code (no multiple-inheritance, unfortunately)
    private String mSrc;

    @Override
    protected void clear()
    {
        mSrc = ".";
        super.clear();
    }

    protected IEventHandler<Event> mSrcChangedEventNotifier = new EventHandler();

    public String getSrc()
    {
        return mSrc;
    }

    public void setSrc(String newSrc) throws MethodParameterIsNullException,
            MethodParameterIsEmptyStringException
    {
        if (newSrc == null)
            throw new MethodParameterIsNullException();
        if (newSrc.length() == 0)
            throw new MethodParameterIsEmptyStringException();
        String prevSrc = mSrc;
        mSrc = newSrc;
        if (mSrc != prevSrc)
            notifyListeners(new SrcChangedEvent(this, mSrc, prevSrc));
    }

    public URI getURI() throws URISyntaxException
    {
        URI uri = null;
        try
        {
            uri = new URI(getSrc()).resolve(getPresentation().getRootURI());
        }
        catch (IsNotInitializedException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        return uri;
    }

    @Override
    public IMedia copyProtected()
    {
        ExternalAudioMedia copy = (ExternalAudioMedia) super.copyProtected();
        try
        {
            URI.create(getSrc()).resolve(getPresentation().getRootURI());
            String destSrc = getPresentation().getRootURI()
                    .relativize(getURI()).toString();
            if (destSrc.length() == 0)
                destSrc = ".";
            try
            {
                copy.setSrc(destSrc);
            }
            catch (MethodParameterIsEmptyStringException e)
            {
                // Should never happen
                throw new RuntimeException("WTF ??!", e);
            }
            catch (MethodParameterIsNullException e)
            {
                // Should never happen
                throw new RuntimeException("WTF ??!", e);
            }
        }
        catch (URISyntaxException e)
        {
            try
            {
                copy.setSrc(getSrc());
            }
            catch (MethodParameterIsEmptyStringException e1)
            {
                // Should never happen
                throw new RuntimeException("WTF ??!", e1);
            }
            catch (MethodParameterIsNullException e2)
            {
                // Should never happen
                throw new RuntimeException("WTF ??!", e2);
            }
        }
        catch (IsNotInitializedException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        try
        {
            copy.setClipBegin(getClipBegin().copy());
            copy.setClipEnd(getClipEnd().copy());
        }
        catch (TimeOffsetIsOutOfBoundsException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        catch (MethodParameterIsNullException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        return copy;
    }

    // END ILocated common code (no multiple-inheritance, unfortunately)
    private ITime mClipBegin;
    private ITime mClipEnd;

    @Override
    public <K extends DataModelChangedEvent> void notifyListeners(K event)
            throws MethodParameterIsNullException
    {
        if (SrcChangedEvent.class.isAssignableFrom(event.getClass()))
        {
            mSrcChangedEventNotifier.notifyListeners(event);
        }
        else
            if (ClipChangedEvent.class.isAssignableFrom(event.getClass()))
            {
                mClipChangedEventNotifier.notifyListeners(event);
            }
        super.notifyListeners(event);
    }

    @Override
    public <K extends DataModelChangedEvent> void registerListener(
            IEventListener<K> listener, Class<K> klass)
            throws MethodParameterIsNullException
    {
        if (SrcChangedEvent.class.isAssignableFrom(klass))
        {
            mSrcChangedEventNotifier.registerListener(listener, klass);
        }
        else
            if (ClipChangedEvent.class.isAssignableFrom(klass))
            {
                mClipChangedEventNotifier.registerListener(listener, klass);
            }
            else
            {
                super.registerListener(listener, klass);
            }
    }

    @Override
    public <K extends DataModelChangedEvent> void unregisterListener(
            IEventListener<K> listener, Class<K> klass)
            throws MethodParameterIsNullException
    {
        if (SrcChangedEvent.class.isAssignableFrom(klass))
        {
            mSrcChangedEventNotifier.unregisterListener(listener, klass);
        }
        else
            if (ClipChangedEvent.class.isAssignableFrom(klass))
            {
                mClipChangedEventNotifier.unregisterListener(listener, klass);
            }
            else
            {
                super.unregisterListener(listener, klass);
            }
    }

    protected IEventHandler<Event> mClipChangedEventNotifier = new EventHandler();

    private void resetClipTimes()
    {
        mClipBegin = new Time().getZero();
        mClipEnd = new Time().getMaxValue();
    }

    protected ExternalAudioMedia()
    {
        super();
        mSrc = ".";
        resetClipTimes();
    }

    @Override
    public boolean isContinuous()
    {
        return true;
    }

    @Override
    public boolean isDiscrete()
    {
        return false;
    }

    @Override
    public boolean isSequence()
    {
        return false;
    }

    @Override
    public ExternalAudioMedia copy()
    {
        return (ExternalAudioMedia) copyProtected();
    }

    @Override
    public ExternalAudioMedia export(Presentation destPres)
            throws MethodParameterIsNullException,
            FactoryCannotCreateTypeException
    {
        if (destPres == null)
        {
            throw new MethodParameterIsNullException();
        }
        return (ExternalAudioMedia) exportProtected(destPres);
    }

    @Override
    protected IMedia exportProtected(Presentation destPres)
            throws MethodParameterIsNullException,
            FactoryCannotCreateTypeException
    {
        if (destPres == null)
        {
            throw new MethodParameterIsNullException();
        }
        ExternalAudioMedia exported = (ExternalAudioMedia) super
                .exportProtected(destPres);
        if (exported == null)
        {
            throw new FactoryCannotCreateTypeException();
        }
        try
        {
            URI.create(getSrc()).resolve(getPresentation().getRootURI());
            String destSrc = destPres.getRootURI().relativize(getURI())
                    .toString();
            if (destSrc.length() == 0)
                destSrc = ".";
            try
            {
                exported.setSrc(destSrc);
            }
            catch (MethodParameterIsEmptyStringException e)
            {
                // Should never happen
                throw new RuntimeException("WTF ??!", e);
            }
        }
        catch (URISyntaxException e)
        {
            try
            {
                exported.setSrc(getSrc());
            }
            catch (MethodParameterIsEmptyStringException e1)
            {
                // Should never happen
                throw new RuntimeException("WTF ??!", e1);
            }
        }
        catch (IsNotInitializedException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        try
        {
            exported.setClipBegin(getClipBegin().copy());
            exported.setClipEnd(getClipEnd().copy());
        }
        catch (TimeOffsetIsOutOfBoundsException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        return exported;
    }

    @Override
    protected void xukInAttributes(IXmlDataReader source, IProgressHandler ph)
            throws MethodParameterIsNullException,
            XukDeserializationFailedException, ProgressCancelledException
    {
        if (source == null)
        {
            throw new MethodParameterIsNullException();
        }
        // To avoid event notification overhead, we bypass this:
        if (false && ph != null && ph.notifyProgress())
        {
            throw new ProgressCancelledException();
        }
        super.xukInAttributes(source, ph);
        String val = source.getAttribute("src");
        if (val == null || val.length() == 0)
            val = ".";
        try
        {
            setSrc(val);
        }
        catch (MethodParameterIsEmptyStringException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        resetClipTimes();
        ITime cbTime, ceTime;
        try
        {
            cbTime = new Time(source.getAttribute("clipBegin"));
            ceTime = new Time(source.getAttribute("clipEnd"));
        }
        catch (MethodParameterIsEmptyStringException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
        catch (TimeStringRepresentationIsInvalidException e)
        {
            throw new XukDeserializationFailedException();
        }
        try
        {
            if (cbTime.isNegativeTimeOffset())
            {
                setClipBegin(cbTime);
                setClipEnd(ceTime);
            }
            else
            {
                setClipEnd(ceTime);
                setClipBegin(cbTime);
            }
        }
        catch (TimeOffsetIsOutOfBoundsException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
    }

    @Override
    protected void xukOutAttributes(IXmlDataWriter destination, URI baseUri,
            IProgressHandler ph) throws MethodParameterIsNullException,
            XukSerializationFailedException, ProgressCancelledException
    {
        if (destination == null || baseUri == null)
        {
            throw new MethodParameterIsNullException();
        }
        if (ph != null && ph.notifyProgress())
        {
            throw new ProgressCancelledException();
        }
        destination.writeAttributeString("clipBegin", this.getClipBegin()
                .toString());
        destination.writeAttributeString("clipEnd", this.getClipEnd()
                .toString());
        if (getSrc() != "")
        {
            URI srcUri;
            try
            {
                srcUri = new URI(getSrc());
            }
            catch (URISyntaxException e)
            {
                throw new XukSerializationFailedException();
            }
            if (baseUri.toString() == "")
            {
                destination.writeAttributeString("src", srcUri.toString());
            }
            else
            {
                destination.writeAttributeString("src", baseUri.relativize(
                        srcUri).toString());
            }
        }
        super.xukOutAttributes(destination, baseUri, ph);
    }

    @Override
    public ITimeDelta getDuration()
    {
        try
        {
            return getClipEnd().getTimeDelta(getClipBegin());
        }
        catch (MethodParameterIsNullException e)
        {
            // Should never happen
            throw new RuntimeException("WTF ??!", e);
        }
    }

    public ITime getClipBegin()
    {
        return mClipBegin;
    }

    public ITime getClipEnd()
    {
        return mClipEnd;
    }

    public void setClipBegin(ITime beginPoint)
            throws MethodParameterIsNullException,
            TimeOffsetIsOutOfBoundsException
    {
        if (beginPoint == null)
        {
            throw new MethodParameterIsNullException();
        }
        if (beginPoint.isLessThan(new Time().getZero()))
        {
            throw new TimeOffsetIsOutOfBoundsException();
        }
        if (beginPoint.isGreaterThan(getClipEnd()))
        {
            throw new TimeOffsetIsOutOfBoundsException();
        }
        if (!mClipBegin.isEqualTo(beginPoint))
        {
            ITime prevCB = getClipBegin();
            mClipBegin = beginPoint.copy();
            notifyListeners(new ClipChangedEvent(this, getClipBegin(),
                    getClipEnd(), prevCB, getClipEnd()));
        }
    }

    public void setClipEnd(ITime endPoint)
            throws MethodParameterIsNullException,
            TimeOffsetIsOutOfBoundsException
    {
        if (endPoint == null)
        {
            throw new MethodParameterIsNullException();
        }
        if (endPoint.isLessThan(getClipBegin()))
        {
            throw new TimeOffsetIsOutOfBoundsException();
        }
        if (!mClipEnd.isEqualTo(endPoint))
        {
            ITime prevCE = getClipEnd();
            mClipEnd = endPoint.copy();
            notifyListeners(new ClipChangedEvent(this, getClipBegin(),
                    getClipEnd(), getClipBegin(), prevCE));
        }
    }

    @Override
    public ExternalAudioMedia splitProtected(ITime splitPoint)
            throws MethodParameterIsNullException,
            TimeOffsetIsOutOfBoundsException
    {
        if (splitPoint == null)
        {
            throw new MethodParameterIsNullException();
        }
        if (splitPoint.isLessThan(getClipBegin()))
        {
            throw new TimeOffsetIsOutOfBoundsException();
        }
        if (splitPoint.isGreaterThan(getClipEnd()))
        {
            throw new TimeOffsetIsOutOfBoundsException();
        }
        ExternalAudioMedia splitAM = copy();
        setClipEnd(splitPoint);
        splitAM.setClipBegin(splitPoint);
        return splitAM;
    }

    @Override
    public boolean ValueEquals(IMedia other)
            throws MethodParameterIsNullException
    {
        if (!super.ValueEquals(other))
            return false;
        ExternalAudioMedia otherAudio = (ExternalAudioMedia) other;
        if (!getClipBegin().isEqualTo(otherAudio.getClipBegin()))
            return false;
        if (!getClipEnd().isEqualTo(otherAudio.getClipEnd()))
            return false;
        if (!(other instanceof ILocated))
        {
            return false;
        }
        try
        {
            if (getURI() != ((ILocated) other).getURI())
                return false;
        }
        catch (URISyntaxException e)
        {
            e.printStackTrace();
            return false;
        }
        return true;
    }

    @Override
    protected void xukOutChildren(IXmlDataWriter destination, URI baseUri,
            IProgressHandler ph) throws XukSerializationFailedException,
            MethodParameterIsNullException, ProgressCancelledException
    {
        /**
         * Does nothing.
         */
    }
}
